### DISCUSSION TITLE
The flash problems of ai decks

### DISCUSSION THREAD

**Discussion by buabula** [bitcraze/discussions#1700] [Q&A] [answered]
Hello,
**Question Description:**
I have two AI Decks. One of them ( referred to as Deck A) previously worked successfully with the WiFi Video Streamer example using a JLink debugger, and I was able to retrieve images. However, when I tried to flash other programs via cfloader, the process got stuck. I decided to reflash the WiFi Video Streamer example, but this time, the computer could no longer detect the WiFi. I attempted to redo the process, but flashing the bootloader failed and got stuck.

Then I switched to another AI Deck (referred to as Deck B) and a different debugger (ARM-USB-TINY-H). Using this new setup, I managed to flash the bootloader on Deck B successfully. However, I still couldn't update the firmware via cfloader.

Below are the detailed steps and results of my attempts:

**Steps to Reproduce:**
Following the tutorial: [Getting Started with AI Deck](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/)

1. Update the CF2.1 and AI Deck firmware using cfclient
Firmware version: 2024.10.2 - firmware-cf2
    
![image](https://github.com/user-attachments/assets/4a4b3077-cb4c-4910-b495-7b5093738a9a)

2. Flash the bootloader with the following command:

```
sudo docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/ftdi/olimex-arm-usb-tiny-h.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/; make all image flash'
```

However, the process got stuck:
![image](https://github.com/user-attachments/assets/ce2e06b1-ca5e-43ba-ab1a-339a066c664c)

I reconnected the battery and the cable to the computer, and then ran the following command, as shown below:
![image](https://github.com/user-attachments/assets/69eed3b1-e798-4816-85eb-6ff34c68972d)

Tried the same process with Deck B:
![image](https://github.com/user-attachments/assets/f263f507-82f1-4d9d-9023-beb5589750a7)
![image](https://github.com/user-attachments/assets/44672388-5045-4532-9237-4399c81e1efd)

While it appeared successful, the next step got stuck again:
![image](https://github.com/user-attachments/assets/9949c87f-6ab9-413a-af6c-dd293249ac26)

**Setup Details:**
AI Deck hardware version: **Rev.C1**
Crazyflie hardware version: **2.1**
no other Decks connected
Platform: **Ubuntu 20.04** (not WSL or VM)

OpenOCD Version:
![image](https://github.com/user-attachments/assets/b212c4e6-2cf4-4516-80d4-6f3143f7f99e)
Debugger:ARM-USB-TINY-H
Configuration Files:
![image](https://github.com/user-attachments/assets/84a945aa-097d-4fa4-8523-40b7dde0c040)
（added ‘adapter speed 100‘）
![image](https://github.com/user-attachments/assets/a26b01a1-98c6-409a-b699-981350ec2297)
(unmodified file）

**Summary of the Issue:**
Both AI Decks encounter flashing issues at different steps. I would greatly appreciate any advice or solutions to resolve these problems.
Thank you so much for your time and support!

**Comment by buabula** [comment]
Additionally, I want to attach two videos for reference: the first shows the hardware while flashing the bootloader using the debugger, and the second shows the hardware while wirelessly flashing via cfloader.

https://github.com/user-attachments/assets/9e262cae-6d1d-47f6-ba77-564f16ed2529


https://github.com/user-attachments/assets/22f6871c-e360-4ff7-b69d-001a1abdb2b8

**Comment by gemenerik** [comment]
Thank you for your post, and I’m sorry to hear you’re having trouble flashing the AI deck.

For future posts, please take a moment to fix the formatting, as some images and a video were not showing, and text was incorrectly highlighted as code or vice versa.

Regarding your issue.  It seems the bootloader flashing was successful on "deck B." Can you verify whether the ESP32 was flashed during the firmware update with the client? After updating the NRF and STM32, the drone should detect the decks, restart, and then flash the ESP32, which can take some time.

I’m not sure what you’re attempting with the OpenOCD commands on "deck A." Have you retried flashing the bootloader using the Olimex debugger on that deck? Ensure everything is properly powered during this process.

If you’re still encountering issues, consider turning the system off for a few minutes. Residual electrical charge in the capacitors can sometimes interfere with the proper initialization of the GAP8.

**Reply by buabula** [reply]
Thank you for your reply, and I am sorry for the formatting errors. I will ensure better formatting in future posts.

I reattempted the flashing process for both decks and found that deck B can now successfully run the WiFi example. However, deck A still gets stuck during the bootloader flashing process. Specifically, while using the same Olimex debugger, I was able to flash the bootloader on deck B without issues, but the process failed on deck A. The current console output for deck A shows the following:
![image](https://github.com/user-attachments/assets/15c87a01-e69a-4ab5-a09a-11f7b7b7420b)

Meanwhile, the console output for deck B displays the following (which may indicate normal behavior):
![image](https://github.com/user-attachments/assets/b3912455-b963-4053-9b41-f03118e9bff1)


The purpose of using the OpenOCD commands was to indicate that the process got stuck when flashing the bootloader for deck A and it didn't flash properly, although the connection between the flasher and the computer should be normal. So the commands were mainly to supplement the flashing problem of deck A.

Thank you again for your suggestions. I look forward to your reply!

**Reply by buabula** [reply]
Hello,

I tried the process once again, but this time I was unable to flash the example via cfloader using deck B. The related video is attached below:

https://github.com/user-attachments/assets/2d36c91d-5d94-4ef4-b45d-cdc564704ec0

Does this mean I need to flash the bootloader every time before flashing the firmware?

**Reply by gemenerik** [reply]
Sorry to hear you are still having problems with flashing. Even though you shouldn't need to reflash the bootloader every time, I wonder; have you tried this, and does it seem to help? 

Have you tried turning off the system a few minutes before flashing? You can also try the resistor workaround mentioned in bitcraze/aideck-gap8-examples#105. Adding a 4.5 kΩ resistor between VCOM and ground can help drain the capacitor properly, ensuring the system shuts off cleanly and allowing the flashing to proceed.

**Reply by buabula** [reply]
Thank you very much for your reply. In fact, a few days after encountering the issue, when I tried to flash the firmware on Deck B, which already had the bootloader flashed, using cfloader, the process got stuck again. So, I was wondering if I need to reflash the bootloader every time.
﻿
At the same time, when I tried to flash the bootloader on Deck A a few days later, it still didn’t work. At that time, I thought I had already “turned off the system for a few minutes before flashing,” since a few days had passed. However, now I realize that perhaps I should have disconnected the power for a few minutes after flashing the firmware via cfclient. I will try this method again.
﻿
Additionally, I am a bit concerned that I may not fully understand the "resistor workaround," so I would like to confirm the exact procedure with you. Should I, or can I, connect a 4.5 kΩ resistor between pin 10 of CRAZYFLIE_EXP_LEFT and pin 1 of U7 to implement your suggestion? I am not sure how to use the battery holder for the solder.
﻿
![image](https://github.com/user-attachments/assets/ae5c6dfe-f153-4947-9cdc-cff3cd48e667)
![image](https://github.com/user-attachments/assets/9f6b2ead-abd5-4ed6-91da-2a3fae4457ac)
![07af704cdb8a79d0c427a7812d5d79e](https://github.com/user-attachments/assets/89093bca-9491-4af7-b125-651819d49718)

Thank you again for your patience! I look forward to your reply.

**Reply by gemenerik** [reply]
You can connect the resistor between the VCOM and GND pins of the [expansion port pinout](https://www.bitcraze.io/documentation/system/platform/cf2-expansiondecks/#expansion-port-pinout). I recommend **against** soldering a resistor directly onto the AI-deck itself; the key is ensuring the resistor connects VCOM and GND at any convenient point in the circuit. 

Instead we recommend using a battery holder. The battery holder includes a pin socket soldered to an otherwise empty PCB, allowing you to solder to the pads supporting the socket. This approach minimizes the risk of damaging an expensive deck or the main PCB that contains sensitive circuitry.

In my case, I used some sort of breakout deck instead, but as you can tell this is not very practical for flying.

![IMG_0382](https://github.com/user-attachments/assets/da6e2002-fff6-4ec2-97b2-4bfd6eceb463)

**Reply by buabula** [reply]
Thank you for your suggestion. Unfortunately, after reattempting the process on both Deck A and Deck B, the issue remains unresolved. For both decks, I followed the procedure of disconnecting the battery, connecting a 4.5 kΩ potentiometer as shown in the image, waiting for five minutes or longer, and then executing the cfloader command, but the wireless flashing process still gets stuck. Additionally, I noticed that the process gets stuck at different points depending on the firmware I’m trying to flash.

Could you kindly provide the steps and commands for flashing the example firmware using the programmer? I would like to try wired flashing on Deck B. Furthermore, Deck A still cannot flash the bootloader via the programmer. Could this issue be related to my pressing the reset button at some point?
![11](https://github.com/user-attachments/assets/ae7c4d28-f77d-4b06-b112-01edbc948ce3)
![1](https://github.com/user-attachments/assets/90a54681-2c67-4b6a-9536-4665812616d6)
![2](https://github.com/user-attachments/assets/3a098d2c-f9be-410b-8d26-461d516479fb)
![3](https://github.com/user-attachments/assets/f63d078a-fd66-43f6-a53c-350f7f687ed7)

Thank you again for your reply!

**Reply by gemenerik** [reply, marked as answer]
Sorry to hear it is still freezing. It freezes at different points because buffers are filling up. If you flash a smaller image, a larger percentage of the image fits into these buffers, hence the difference in percentage where it freezes.

Can you make sure the battery is fully charged. And keep the resistor attached. Then run through all the steps of the ["Preparing the AI-deck and Crazyflie firmware" section](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#preparing-the-ai-deck-and-crazyflie-firmware) of the getting started guide (again). So updating the CF and AI-deck through the client, flashing the GAP8 bootloader, and finally flashing the WiFi example.

**Reply by buabula** [reply]
Thank you very much for your reply! Following your suggestion, I connected the resistor from the beginning of the flashing process, and it worked. Both Deck A and Deck B can now be successfully flashed via `cfloader`. Although Deck A still gets stuck when flashing the bootloader through the programmer, it seems possible that since I successfully flashed the bootloader previously, this current failure might not impact the firmware flashing process.

Now that I can flash the AI Deck firmware wirelessly, the most critical issue has been resolved. Thank you so much for your help!  

And should I open a separate issue to address the bootloader flashing problem?

**Reply by gemenerik** [reply]
Happy to hear it's working now! Yes, please open a new thread for the bootloader flashing problem, with details about how everything is connected, powered, what commands you use etc.
